src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKCommandLinePluginDescriptor.java

```diff
@@ -0,0 +1,142 @@
+ * command line parser and specified as command line arguments.
+ *
+ * Descriptors (sub)classes:
+ *
+ * - must live in the org.broadinstitute.hellbender.cmdline.GATKPlugin package
+ * - must have a no-arg constructor
+ * - should have at least one @Argument used to accumulate the user-specified instances
+ *   of the plugin seen on the command line. Allowed values for this argument are determined
+ *   by the descriptor, but are usually the simple class names of the discovered plugin
+ *   subclasses.
+ *
+ * Each descriptor passed to the command line parser is instantiated before argument
+ * parsing starts, and is queried to find the command line plugin classes that should be
+ * discovered dynamically.
+ *
+ * Plugin classes:
+ *
+ * - should subclass a common base class (the name of which is returned by the descriptor)
+ * - should all live in the same package (to allow them to be unambiguously specified
+ *   by the class' simple name
+ * - should contain @Arguments for any values they wish to collect. @Arguments may be
+ *   optional or required. If required, the arguments are in effect "provisionally
+ *   required" in that they are contingent on the specific plugin being specified on
+ *   the command line; they will only be marked by the command line parser as missing
+ *   if the they have not been specified on the command line, and the plugin class
+ *   containing the plugin argument *has* been specified on the command line (as
+ *   determined by the command line parser via a call to isDependentArgumentAllowed).
+ *
+ * NOTE: plugin class @Arguments should not have initial values, as the command line
+ * parser will interpret these as having been set (and will validate them accordingly
+ * and throw if the dependent argument has not been set).
+ *
+ * The methods for each descriptor are called in the following order:
+ *
+ *  getPluginClass()/getPackageName() - once when argument parsing begins (if the descriptor
+ *  has been passed to the command line parser as a target descriptor)
+ *
+ *  getClassFilter() - once for each plugin subclass found
+ *  addInstance() - once for each plugin subclass that isn't filtered out by getClassFilter
+ *  validateDependentArgumentAllowed  - once for each plugin argument value that has been
+ *  specified on the command line for a plugin that is controlled by this descriptor
+ *
+ *  validateArguments() - once when argument parsing is complete
+ *  getInstances() - whenever the pluggable class consumer wants the resulting plugin instances
+ */
+public abstract class GATKCommandLinePluginDescriptor<T> {
+
+    /**
+     * Base class to be used as command line plugin. Subclasses of this class will be
+     * command line accessible.
+     */
+    public abstract Class<?> getPluginClass();
+
+    /**
+     * Package names from which to load the command line plugin classes
+      */
+    public abstract String getPackageName();
```

Comment:
Can we make this return a `List<String>`, to allow future support for multiple packages per plugin type?